### DISCUSSION TITLE
enabling & standardize automatic, incremental Index updates

### DISCUSSION THREAD

**Discussion by hpvd** [duckdb/duckdb#15291] [Ideas]
This is somehow a superset of and an extract from:
- Feature request: Incrementally updating FTS index #5352
- Feature Request: Make FTS index update on INSERT, UPDATE, DELETE #15094

----
**Enabling & standardize automatic, incremental Index updates**
Lots of the speed possibilities of DBs results from indexes, so they should have to be used as often as possible...
To make this possible, they have to be up-to-date, which is not really possible in many use-cases without relying on automatic and incremental updates...

Just to give 2 examples what indexes can do:
1. with index active, classic relational dbs (MariaDB/Mysql) can even outperform graph dbs (neo4j) in typical graph use-cases using queries incl `with recursive` e.g. see task 5 in https://www.mdpi.com/2076-3417/12/13/6490
(duckdb has a good chance to do the same, it already supports `with recursive` too  https://duckdb.org/docs/sql/query_syntax/with.html)

3. In the OLAP space, Pinot is the golden sample to benefit from Indexes in a stunning way:
see many benchmarks with and without indexes:
https://startree.ai/resources/what-makes-apache-pinot-fast-chapter-ii

Two of some examples from this site:

both from a data-set with **3 billion rows**:

A) Count the number of events for a given actor_id: with and without sorted index
![2025-01-08_11h12_33](https://github.com/user-attachments/assets/a4dc3aa0-7183-4aef-bbdb-489d32c2cf80)

B) Count the number of events for pull requests with ‘pinot’ in the title: with and without text index
![2025-01-08_11h13_59](https://github.com/user-attachments/assets/f8f5d28a-bf25-4435-9d3a-1fe9c85d577c)

=> Maybe one day one can/should provide
- the needed mechanism incl triggers on update, insert, delete and
- some kind of architecture guide

to implement automatic and incremental updates for any kind of index.

What do you think?

**Comment by hpvd** [comment]
Currently we may not only leave additional speed beyond the already fantastic progress of Duckdb over the years
https://duckdb.org/2024/06/26/benchmarks-over-time.html
and possible production OLAP use-cases for the 99% who don't need real big data beyond a single machine https://motherduck.com/blog/big-data-is-dead/
but also people start working around e.g. in the field of Full Text Search https://github.com/duckdb/duckdb/discussions/5352#discussioncomment-11534624

**Comment by lnkuiper** [comment]
I'm not sure what you mean by this:

> to implement automatic and incremental updates for any kind of index.

DuckDB already supports incremental updates for the index that is used in `CREATE INDEX`. The FTS index is a "special" kind of index, as it is _not an index in the traditional SQL sense_ (such as indexes used to enforce constraints like `UNIQUE`). Rebuilding this index on every update would not make any sense.

Just to clarify:

Instead of a traditional index, the FTS extension is just a bunch of SQL queries to construct and query a full-text-search index. If you have a table `documents` in schema `main`, you will have a schema `fts_main_documents` that holds the index. It is essentially a bunch of tables. For this kind of "index", our usual index updates cannot work.

To support incremental updates for this "bunch of tables", we would need some kind of `TRIGGER`, as we just want to update some other tables when the input table gets updated. These updates can also be expressed purely in SQL.

**Reply by hpvd** [reply]
> Instead of a traditional index, the FTS extension is just a bunch of SQL queries to construct and query a full-text-search index. If you have a table `documents` in schema `main`, you will have a schema `fts_main_documents` that holds the index. It is essentially a bunch of tables. For this kind of "index", our usual index updates cannot work.

ah ok, I wasn't aware of this.
The idea behind this topic is to look for ways/structures to make it possible to have **all indexes**, even the new ones brought by extensions
like
- for fulltext search,
- for spatical data,
- for vector similarity
- ...

being able to be updated in automatic fashion and in incremental way.
Of course these indexes have to been build a suitable way and need some kind of "tiny trigger api" (therefore was the named architecture guide).

So its somehow about standardisation of index behaviour, to always take the most advantage of them.

**Reply by lnkuiper** [reply]
The spatial/vss indexes are actual "SQL indexes" and are updated automatically, AFAIK:
1. https://duckdb.org/docs/extensions/vss.html#inserts-updates-deletes-and-re-compaction
2. https://duckdb.org/docs/extensions/spatial/r-tree_indexes#bulk-loading--maintenance

While triggers are nice (and generally useful), they will most likely not apply to any indexes created using `CREATE INDEX`, as these are SQL indexes and are updated automatically.

The FTS index is not updated automatically because it is not an "SQL index" created using `CREATE INDEX` (and it can't be—indexing and retrieval requires creating and querying tables!).

So, while I agree that we would like to have triggers, it is not because they would enable updating SQL indexes—these are already updated. They would allow some user-defined automation, including updating the FTS index.

**Reply by hpvd** [reply]
thanks for insights and even further clarifying!

**Comment by hpvd** [comment]
just added two examples
